### ISSUE TITLE
Discussion: Blackoil API

### ISSUE THREAD

**Issue by andlaus** [OPM/opm-material#101]
this issue is intended to be the continuation of the conversation of #98 and is thus some kind of a TODO list. So far we (@atgeirr and me) have agreed on the following things (If I forgot something, or we did not actually agree, please add a comment):

(Edit from @atgeirr: I'll edit directly in your list just this time, to avoid massive quoting. See the "Edited:" items below. It can be assumed that I agree with what it says now. )

- the API in opm-material stays "single point" and won't switch to "bulk mode"
  - some convenience layer may be added to the simulator modules which use a "global" POV. (I suppose this will only be done if the wrapper is actually more convenient to use than explicit loops. Let's see.)
- The API for the low-level PVT objects for gas, oil and water stays largely unchanged:
  - The parameters passed to these objects methods stay as they are
  - Unifying classes will only be done if the result is an improvement on the current state. (IMO this mean "clarity of the code". Edited: @atgeirr also thinks that changes which only result in fewer classes are worth it. possibly these two things are not mutually exclusive, though)
  - renaming some methods would be beneficial (but I forgot which these were, please add comments). Edited: the method dealing with the saturation curve should use the word "saturated" to make more clear what they do.
- the API of the BlackOilFluidSystem will become more consistent and higher-level:
  - instead of explicitly passing the quantities on which a method depends on, a "fluid state" object is passed everywhere
  - the phase/component indices are passed as method arguments (i.e., they will no longer encoded into the method names)
- the mapping from active to canonical phase indices will continue to happen outside of the fluid system
  - the initialization code needs to deal with error conditions
- all methods of BlackOilFluidSystem get black-oil specific variants:
  - instead of via the `ParameterCache` object, the PVT region index is passed explicitly as an integer to these
  - the `ParameterCache` taking variants of these methods are simple forwards to the new methods
- The fact that the computation methods are static is unfortunate:
  - it makes instantiation of multiple fluid systems (e.g., for comparison purposes) much harder than it needs to be
  - it is a lot of work to change the approach to the normal "instanced classes" pattern because every single model in eWoms would need a lot of changes as would most of the constraint solvers from opm-material
  - in retrospect, it thus would have been better to use instanced classes, and we might change to that pattern some time in the future, but it is not on the immediate agenda
- some methods of the BlackOilFluidSystem must be renamed for consistency
- Edited: the phase parameters required should be strongly typed (new enum class perhaps?) and not convertible from int for safety.

**Comment by andlaus** [comment]
> Edited: the phase parameters required should be strongly typed (new enum class perhaps?) and not convertible from int for safety.

here I have a question: how to do loops over all phases or components in this scenario? I consider this to be an important issue because a lot (most) of code actually does not really care what a given index means, but just does stuff like

```c++
Scalar totalMobility = 0.0;
for (int phaseIdx = 0; phaseIdx < numPhases; ++phaseIdx)
   totalMobility += kr[phaseIdx]/FluidSystem::viscosity(phaseIdx);
```

(slight pseudo-code)

**Comment by atgeirr** [comment]
> here I have a question: how to do loops over all phases or components in this scenario?

I thought about that. If we have
```
enum class Phase = { Water, Oil, Gas }; // Names can of course be discussed
```
and we need to call a function declared like this:
```
template <class State>
auto viscosity(Phase phase, const State& state) -> decltype(State::Scalar);
```
in a loop, it is fairly simple:
```
std::vector<Phase> active_phases = { Phase::Oil, Phase::Gas }; // For example. Initialise based on data.
...
for (Phase phase : active_phases) { // Or loop with integer index p and use active_phases[p].
    ...
    auto visc = fluidsystem.viscosity(phase, state);
}
```

**Comment by andlaus** [comment]
> in a loop, it is fairly simple:

I suppose that deciding whether what you've suggested or
```c++
for (int phaseIdx = 0; phaseIdx < FluidSystem::numPhases; ++phaseIdx) {
[...]
}
```
is simpler is in the eye of the beholder ;)

I have more serious issues with this approach, though:
- it assumes that the *names* of the phases are canonical for all fluid systems. in general, this cannot be the case: some fluid systems from remediation applications may want to call the oil phase "NAPL" or others from chemical engineering may have e.g., acetone instead of water. (note that the "canonical phase name" approach completely breaks down for twophase fluid systems because for these each phase could be either of the liquids or the gas). AFAICS, the canonical phase names would also not be "oil", "water" and "gas", but rather "polar liquid", "apolar liquid" and "gas"?
- your approach pushes people do some seriously dangerous patterns ("dangerous" == "slow"): `std::vector` allocates some memory on the heap every time it is allocated. this potentially leads to bad performance if you do "point-wise" calculations. I know that this can be avoided using `std::array`, but as you showed yourself, people will easily get this wrong. (and even with `std::array` you'll probably interfere with the compiler's constant expression elimination.)

Note that I'd very much like to make phase and component indices type safe, but these types must (a) support `operator++()`, (b) not come with any runtime performance impact and (c) should not be too much hassle to implement for both, the users of the API as well as for its implementor.

**Comment by atgeirr** [comment]
To your first point about
```
for (int phaseIdx = 0; phaseIdx < FluidSystem::numPhases; ++phaseIdx) {
[...]
}
```
being just as simple: in my opinion, the above is clearly not as simple, because there is ambiguity about active phases vs. canonical phases, especially for new users who would not have read our discussion.

To you second point (first bullet), it is indeed true that names vary, but I do not insist that the names must be canonical across all fluid systems, indeed I guess that this is one of the ways fluid systems could vary. We are discussing the BlackoilFluidSystem, and for that system the phase names as suggested (or similar) make sense, and are indeed already embedded in (current at least) naming of methods!

To your third point (second bullet), such a std::vector should not be created locally (I just did to show the idea), but should live as a member in a higher-level fluid or model class, and be initialized depending on the actual phases used for the run. I agree there could be a small pitfall here, and we should make sure to avoid this. I do not think it is huge though.

To your fourth point: a) Why do you want the increment operator? Oil++ does not make sense to me. Iteration over all active phases must be done with knowledge of which phases are active, anyway. b) I agree there should be no runtime penalty, and believe there is none in my suggestion (assuming no abuse of local vectors, but I find that to be an unlikely mistake). c) I do believe it is not much of a hassle either, can you point out where you think this would be bad?

**Comment by andlaus** [comment]
> To you second point (first bullet), it is indeed true that names vary, but I do not insist that the names must be canonical across all fluid systems, indeed I guess that this is one of the ways fluid systems could vary. 

but in your scenario, not requiring canonical names for all fluid systems would make writing generic code (which does not care about the fluid system or the names of phase indices) impossible, right?

> To your third point (second bullet), such a std::vector should not be created locally (I just did to show the idea), 

yeah, I see this. my point was not primarily that this cannot be done right, the point was that what you suggest makes doing the wrong thing very easy. (maybe even easier than doing it correctly?)

> a) Why do you want the increment operator?

because I want an easy and performant way loop over all phases? In principle, something like

```c++
struct BlackOilPhaseIndex {
  enum PhaseIndexEnum { gasPhaseIdx, oilPhaseIdx, waterPhaseIdx };
  PhaseIndexEnum phaseIdxEnum;
  BlackOilPhaseIndex& operator++(int) { phaseIdxEnum = static_cast<PhaseIndexEnum>(++ static_cast<int>(phaseIdxEnum)); return *this; };
  // [...]
};

// [...]
class BlackOilFluidSystem {
  // [..]
  typedef BlackOilPhaseIndex PhaseIndex;
  // [...]
}
```

would probably do the trick. I think that it may be too little return for the investment, though. (and it may trigger the compiler to do some additional padding, but maybe it doesn't.)

> `Oil++` does not make sense to me.

right. but `++ phaseIdx` makes a lot of sense.

> Iteration over all active phases must be done with knowledge of which phases are active, anyway. 

here comes the philosophical difference again: IMO it is a bad mistake to embed the concept of "active phases" into the fluid system: The fluid system should just provide the most general case and if the simulator decides that a phase is inactive, it should simply not query for its quantities.

I propose to defer this discussion until the PRs for converting opm-{core,autodiffs} have been opened.

> c) I do believe it is not much of a hassle either, can you point out where you think this would be bad?

try to port the `NcpFlash` class (which is included in opm-material). I suppose that you'll see what I mean as soon as you feed it any non-blackoil fluid system.

**Comment by atgeirr** [comment]
> because I want an easy and performant way loop over all phases? In principle, something like
> (removed)
> would probably do the trick. I think that it may be too little return for the investment, though. (and it may trigger the compiler to do some additional padding, but maybe it doesn't.)

That does seem excessive, yes.

>> Oil++ does not make sense to me.

> right. but ++ phaseIdx makes a lot of sense.

So we need to separate the phase index from the phase ID. My wish is to use the phase IDs in the API and not the indices. As my example shows it is still easy to iterate over all phases. However, for a more generic code where the only appropriate thing to do is to iterate over all phases, without any knowledge of phase names etc., one could provide a method allPossiblePhases(), that returns some container of all the phase IDs the system accepts. (I would then strongly suggest that activePhases() is provided as well, I think the fluid system should not allow gas queries if there is no gas phase, providing default answers is not a good idea).

> 
> Iteration over all active phases must be done with knowledge of which phases are active, anyway.
> here comes the philosophical difference again: IMO it is a bad mistake to embed the concept of "active phases" into the fluid system: The fluid system should just provide the most general case and if the simulator decides that a phase is inactive, it should simply not query for its quantities.

I can accept this, if we can agree on the use of (strongly typed) phase IDs for the API, also see above. I think the fluid system is exactly the part of the simulator that knows best which phases are active.

> 
> I propose to defer this discussion until the PRs for converting opm-{core,autodiffs} have been opened.

I'd prefer to have a consensus before we start implementing too much.

>> c) I do believe it is not much of a hassle either, can you point out where you think this would be bad?
> try to port the NcpFlash class (which is included in opm-material). I suppose that you'll see what I mean as soon as you feed it any non-blackoil fluid system.

I may take a look at this, but not having much of an idea of what it does it may not be too fruitful...

Another idea: one can provide a simple wrapper that takes the BlackoilFluidSystem class and provide a generic (canonical index-based) interface, then that wrapper can be used by more generic code (although I like the allPossiblePhases() idea better).

**Comment by andlaus** [comment]
> I may take a look at this, but not having much of an idea of what it does it may not be too fruitful...

oh, that's simple: given the total molarities c^\kappa of the components (defined as 
`
c^\kappa = \sum_\alpha x_\alpha^\kappa c_\alpha
`) and temperature, it computes the full thermodynamic state. (i.e., phase pressures, phase saturations and phase compositions)

**Comment by andlaus** [comment]
I guess that this issue has been largely resolved with #114 and OPM/opm-autodiff#576. closing for now, but if you don't agree, feel free to re-open it.
